### PR TITLE
cw20-merkle-airdrop add possibility to burn remaining tokens after expiration

### PR DESCRIPTION
**Pull request by dimiandre**
After expire time owner can burn the remaining tokens in the contract

(work in progress)

### CODE REVIEW

**Comment by dimiandre** [comment]
Maybe it's better to check if all the stage are expired instead of only the last one

File: contracts/cw20-merkle-airdrop/src/contract.rs [hunk 1]
```diff
@@ -212,6 +213,56 @@ pub fn execute_claim(
...
+    }
+
+    // make sure is expired
+    let expiration = STAGE_EXPIRATION.load(deps.storage, stage)?;
+    if !expiration.is_expired(&env.block) {
+        return Err(ContractError::StageNotExpired { stage, expiration });
+    }
+
+    // Get balance
+    let balance: BalanceResponse = deps
```
Comment (orkunkl) [inline contracts/cw20-merkle-airdrop/src/contract.rs:236]: If multiple stages are going on at the same time, this code will result in wiping out of full airdrop balance 💣 Best to save stage airdrop amount on register_merkle_root phase and burn that amount

**Review by orkunkl — CHANGES_REQUESTED** [review]
Good stuff, just need one change

File: contracts/cw20-merkle-airdrop/src/contract.rs (hunk 1 above)
Comment (dimiandre) [reply contracts/cw20-merkle-airdrop/src/contract.rs]: I had to add also a "total claimed per stage" that updates every time an user claim, but should be ok now

File: contracts/cw20-merkle-airdrop/src/state.rs [hunk 2]
```diff
@@ -24,8 +26,17 @@ pub const STAGE_EXPIRATION: Map<u8, Expiration> = Map::new(STAGE_EXPIRATION_KEY)
...
+pub const STAGE_AMOUNT_CLAIMED_KEY: &str = "claimed_amount";
+pub const STAGE_AMOUNT_CLAIMED: Map<u8, Uint128> = Map::new(STAGE_AMOUNT_CLAIMED_KEY);
+
 pub const MERKLE_ROOT_PREFIX: &str = "merkle_root";
 pub const MERKLE_ROOT: Map<u8, String> = Map::new(MERKLE_ROOT_PREFIX);
 
 pub const CLAIM_PREFIX: &str = "claim";
 pub const CLAIM: Map<(&Addr, u8), bool> = Map::new(CLAIM_PREFIX);
+
+pub const CLAIMED_AMOUNT_PREFIX: &str = "claimed_amount";
```
Comment (orkunkl) [inline contracts/cw20-merkle-airdrop/src/state.rs:41]: Oops this is going to 🤯. Using same key for two states. Also naming is confusing.

**Comment by orkunkl** [comment]
We need tests for these highly critical new features.
Error above shows the need for it.

File: contracts/cw20-merkle-airdrop/src/state.rs (hunk 2 above)
Comment (dimiandre) [reply contracts/cw20-merkle-airdrop/src/state.rs:41]: 🤯 The tests were also passing anyways, very weird. Fixed and added more

**Comment by dimiandre** [comment]
Added more tests, fixed the state key

File: contracts/cw20-merkle-airdrop/src/error.rs [hunk 3]
```diff
@@ -32,6 +32,9 @@ pub enum ContractError {
     #[error("Airdrop stage {stage} expired at {expiration}")]
     StageExpired { stage: u8, expiration: Expiration },
 
+    #[error("Airdrop stage {stage} not expired yet")]
```
Comment (orkunkl) [inline contracts/cw20-merkle-airdrop/src/error.rs:35]:
```suggestion
    #[error("Airdrop stage {stage} expires at {expiration}")]
```
You may need to implement `Display` trait.

**Review by orkunkl — APPROVED** [review]
Added a few optional improvements. Good work!

**Pull request merged**